server/src/main/java/io/crate/execution/engine/aggregation/impl/NullAwareBinaryDocValueAggregator.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Licensed to Crate under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.execution.engine.aggregation.impl;
+
+import org.elasticsearch.common.CheckedBiConsumer;
+import org.elasticsearch.index.fielddata.SortedBinaryDocValues;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.util.function.Supplier;
+
+
+public class NullAwareBinaryDocValueAggregator<T> extends BinaryDocValueAggregator<T> {
```

Comment:
Maybe it deserves a comment that points out that this is like `BinaryDocValueAggregator`, except that it also calls the `docValuesConsumer` if there is no document match.

Took me a bit to figure that out.

I also wonder if we should implement `DocValueAggregator` directly instead of extending `BinaryDocValueAggregator` and repeat a bit of the implementation, to have a smaller hierachy.

Reply:
These `NullAware...` classes are used only in `ArrayAgg`, it makes sense just to implement the Binary and Numeric doc values aggregators directly in `ArrayAgg` and override `apply` a bit.